File: frontend/web/src/api/mentorDistribution.ts
/*
 * GET /api/mentor/distribution — агрегат «Незапущенные команды» по всем
 * проектам ментора. Каждый проект содержит pool заявок (qualified/unqualified
 * × приоритет) и список незапущенных команд с уже-распределёнными
 * участниками. Контракт зеркалит models.MentorDistributionResponse в Go-бэке.
 */

import { apiFetch } from './client';
import type {
  ApplicantPriorityBuckets,
  ApplicantRequirements,
  ApplicationStatus,
} from './applications';
import type { ProjectStatus } from './projects';

export interface MentorDistributionTeamMember {
  applicationId: number;
  studentId: number;
  firstName: string;
  lastName: string;
  course: number;
  group: string;
  gpa: number;
  priority: number;
  status: ApplicationStatus;
  qualified: boolean;
  /** Заполняется только в coord distribution endpoint (mentor не использует). */
  allPriorities?: TeamMemberPriority[];
}

export interface TeamMemberPriority {
  applicationId: number;
  projectId: number;
  projectTitle: string;
  company?: string;
  mentorName?: string;
  priority: number;
  status: ApplicationStatus;
}

export interface MentorDistributionTeam {
  id: number;
  name: string;
  launched: boolean;
  members: MentorDistributionTeamMember[];
}

export interface MentorDistributionPool {
  qualified: ApplicantPriorityBuckets;
  unqualified: ApplicantPriorityBuckets;
}

export interface MentorDistributionProject {
  id: number;
  title: string;
  status: ProjectStatus;
  company: string;
  teamSizeMin: number;
  teamSizeMax: number;
  numTeams: number;
  sprintsCount: number;
  sprintWeeks: number;
  /** ISO YYYY-MM-DD; последняя end_date из спринтов проекта. Может быть пустой. */
  deadline?: string;
  requirements: ApplicantRequirements;
  teams: MentorDistributionTeam[];
  pool: MentorDistributionPool;
}

export interface MentorDistributionResponse {
  projects: MentorDistributionProject[];
}

export function getMentorDistribution(): Promise<MentorDistributionResponse> {
  return apiFetch<MentorDistributionResponse>('/mentor/distribution');
}

File: backend/project-service/internal/models/mentor_distribution.go
package models

// MentorDistributionResponse — агрегированный DTO для GET /api/mentor/distribution.
// Возвращает все проекты ментора, у которых есть незапущенные команды (launched=false)
// или непринятые заявки. Каждый проект содержит мета, команды и пул заявок.
//
// Используется страницей mentor.html view-distribution: drag&drop студентов между
// пулом и слотами незапущенных команд + кнопка «Запустить команду».
type MentorDistributionResponse struct {
	Projects []MentorDistributionProject `json:"projects"`
}

type MentorDistributionProject struct {
	ID                 int                              `json:"id"`
	Title              string                           `json:"title"`
	Status             ProjectStatus                    `json:"status"`
	Company            string                           `json:"company"`
	TeamSizeMin        int                              `json:"teamSizeMin"`
	TeamSizeMax        int                              `json:"teamSizeMax"`
	NumTeams           int                              `json:"numTeams"`
	SprintsCount       int                              `json:"sprintsCount"`
	SprintWeeks        int                              `json:"sprintWeeks"`
	Deadline           string                           `json:"deadline,omitempty"`
	Requirements       ApplicantRequirements            `json:"requirements"`
	Teams              []MentorDistributionTeam         `json:"teams"`
	Pool               MentorDistributionPool           `json:"pool"`
}

type MentorDistributionTeam struct {
	ID       int                            `json:"id"`
	Name     string                         `json:"name"`
	Launched bool                           `json:"launched"`
	Members  []MentorDistributionTeamMember `json:"members"`
}

type MentorDistributionTeamMember struct {
	ApplicationID int               `json:"applicationId"`
	StudentID     int               `json:"studentId"`
	FirstName     string            `json:"firstName"`
	LastName      string            `json:"lastName"`
	Course        int               `json:"course"`
	Group         string            `json:"group"`
	GPA           float64           `json:"gpa"`
	Priority      int               `json:"priority"`
	Status        ApplicationStatus `json:"status"`
	Qualified     bool              `json:"qualified"`
	// AllPriorities — все заявки этого студента (нужно для drawer'а в admin
	// distribution: видеть, какие проекты он указывал в каком приоритете).
	// На менторском distribution не заполняется (omitempty).
	AllPriorities []TeamMemberPriority `json:"allPriorities,omitempty"`
}

// TeamMemberPriority — лёгкая запись о другой заявке этого студента.
type TeamMemberPriority struct {
	ApplicationID int               `json:"applicationId"`
	ProjectID     int               `json:"projectId"`
	ProjectTitle  string            `json:"projectTitle"`
	Priority      int               `json:"priority"`
	Status        ApplicationStatus `json:"status"`
}

type MentorDistributionPool struct {
	Qualified   ApplicantPriorityBuckets `json:"qualified"`
	Unqualified ApplicantPriorityBuckets `json:"unqualified"`
}
